java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java

```diff
@@ -131,16 +137,31 @@ public BlockVisitor(Symbol catchedException) {
 
     @Override
     public void visitMethodInvocation(MethodInvocationTree tree) {
-      Symbol symbol = tree.symbol();
-      if (!symbol.isUnknown() && symbol.owner().type().isSubtypeOf("java.lang.Thread") && "interrupt".equals(symbol.name()) && tree.arguments().isEmpty()) {
+      if (threadInterrupted || depth >= MAX_DEPTH) {
+        return;
+      }
+      depth++;
+      if (INTERRUPT_MATCHERS.matches(tree)) {
         threadInterrupted = true;
         return;
       }
+      Tree declaration = tree.symbol().declaration();
+      if (declaration != null) {
+        //Declaration of MethodInvocationTree is MethodTree
+        BlockTree block = ((MethodTree) declaration).block();
+        if (block != null) {
+          block.accept(this);
+        }
+      }
+      depth--;
       super.visitMethodInvocation(tree);
     }
 
     @Override
     public void visitThrowStatement(ThrowStatementTree tree) {
+      if (threadInterrupted || depth >= 3) {
```

Comment:
`depth` should be compared against `MAX_DEPTH`, not a hard coded `3`, but I don't think it should be compared here at all (see other comment).